### PR TITLE
New VFS

### PR DESCRIPTION
**Pull request by cole-miller**
Very WIP.

### CODE REVIEW

File: src/vfs2.c [hunk 1]
```diff
@@ -0,0 +1,1144 @@
...
+			xfile->db_shm.shared[i] = 0;
+			xfile->db_shm.exclusive[i] = 0;
+		}
+	}
+	return SQLITE_OK;
+}
+
+/* sqlite3_io_methods implementations end here */
+
+static struct sqlite3_io_methods vfs2_io_methods = {3,
```
Comment (just-now) [inline src/vfs2.c:701]: Curious, why don't we use c99 named initializers here. Exaple:
```
tatic struct sqlite3_io_methods vfs2_io_methods = {
 .iVersion = 3,
 .xClose = vfs2_close,
 ...
 }
 ```

**Review by just-now — COMMENTED** [review]
This is the first pass. Will continue reading the logic.

File: src/vfs2.c [hunk 2]
```diff
@@ -0,0 +1,1144 @@
...
+		uint32_t nBackfillAttempted;
+		uint8_t unused[4];
+	} hdr;
+	char bytes[VFS2_WAL_INDEX_REGION_SIZE];
+};
+
+struct vfs2_wal
+{
+	const char *moving_name;   /* e.g. /path/to/my.db-wal */
+	char *wal_cur_fixed_name;  /* e.g. /path/to/my.db-xwal1 */
```
[the remark this answers is not in the document]
Comment (cole-miller) [reply src/vfs2.c:88]: They are "owned" and need to be freed at the end of the file object's lifetime, unlike `moving_name` which is "borrowed" from SQLite (but guaranteed to be valid until xClose, see https://sqlite.org/c3ref/vfs.html. That's why I've left the distinction here

File: src/vfs2.c [hunk 3]
```diff
@@ -0,0 +1,1144 @@
...
+		    QUEUE__DATA(q, struct vfs2_db_entry, queue);
+		if ((f->flags & SQLITE_OPEN_MAIN_DB) && cur->wal != NULL) {
+			const char *walname = cur->wal->wal.moving_name;
+			const char *dash = strrchr(walname, '-');
+			assert(dash != NULL);
+			if (strncmp(walname, f->db_shm.name,
+				    (size_t)(dash - walname)) != 0) {
+				continue;
+			}
+			found = true;
```
[the remark this answers is not in the document]
Comment (cole-miller) [reply src/vfs2.c:210]: The calls are a bit different, it's

strncmp(cur->wal->wal.moving_name, f->db_shm.name, ...)

vs

strncmp(f->wal.moving_name, cur->db->db_shm.name, ...)

I can't see how to factor out what's common without making the code more complicated

**Comment by cole-miller** [comment]
I'm going to merge this into dqlite-next and we can iterate on it there!

**Pull request merged**